<html>
<head>
    <meta charset="utf-8"/>
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" />
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-172722632-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-172722632-1');
    </script>

    <style>
        body, .ui.form, label {
            font-size: 11pt !important;
        }
        td {
            padding: 5px 15px;
        }
        .upperBorder {
            border-top: 1px solid black;
        }
        table {
            border-collapse: collapse;
        }
    </style>
</head>
<body>
    <div class="ui stackable grid">
        <div class="row">
        <div class="one wide column"></div>
        <div class="seven wide column">
            <div style="padding:25px 0px">
                Enter a variant below to see its <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">SpliceAI</a> scores.
                This can be any SNV or simple InDel within the exon or intron of a gene as defined by GENCODE v24 canonical transcripts.<br />
                <br />
                More information on SpliceAI can be found in <a href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et al, Cell 2019</a> and <a href="https://github.com/Illumina/SpliceAI" target="_blank">github.com/Illumina/SpliceAI</a>, <br />
		as well as in this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded talk</a> by the first author of SpliceAI.<br />
                To post issues or feature requests for this web interface, please go to <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">broadinstitute/SpliceAI-lookup/issues</a>.<br />
                <br />
                This web interface works by retrieving scores from an <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">API</a> maintained by the
                <a href="https://the-tgg.org/" target="_blank">TGG</a>.<br />
                The <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">API</a> uses precomputed
                scores provided by Illumina for SNVs and small InDels, and runs the <a href="https://github.com/Illumina/SpliceAI" target="_blank">SpliceAI model</a>
                directly to compute scores for any other queries - such as larger InDels or non-default "Max distance" values.<br/>
                <br />
                Examples: &nbsp; (on hg38) <br />
                <br />
                <div style="padding-left: 30px; color: gray">
                    chr8-140300616-T-G  &nbsp; &nbsp; (<i>the</i> chr <i>prefix is optional</i>)<br />
                    8:140300616 T&gt;G<br />
                    chr8 &nbsp; 140300616 &nbsp; T &nbsp; G<br />
		    1:930130:C:G  (<i>position with overlapping genes</i>)<br />
                    1-1042601-A-AGAGAG  (<i>insertion of GAGAG</i>) <br />
                    1-1042466-GGGC-G  (<i>deletion of GGC</i>) <br />
                    NM_000552.4(VWF):c.3797C>A (p.Pro1266Gln)<br />
                </div >
            </div>
            <div class="ui form" style="width:100%">
                <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                    <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                </div>
                <!-- div class="field" style="padding-bottom: 10px;">
                    <textarea id="search-box2" rows="1" style="width: 100%" placeholder="Enter variant..."></textarea>
                </div -->
                <div class="ui stackable grid">
                    <div class="row">
                        <div class="thirteen wide column">
                            <div class="field" style="padding-right:10px">
                                <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                    <input type="radio" name="hg" value="37" /><label>hg19</label>
                                </div>
                                <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                    <input type="radio" name="hg" value="38" checked /><label>hg38</label>
                                </div>
                            </div>
                            <div class="field" style="padding-right:10px">
                                <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Score type:</b></span>
                                <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                    <input type="radio" name="mask" value="1" /><label>masked</label>
                                </div>
                                <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                    <input type="radio" name="mask" value="0" checked /><label>raw</label>
                                </div>
                                <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Splicing changes corresponding to strengthening annotated splice sites and weakening unannotated splice sites are typically much less pathogenic than weakening annotated splice sites and strengthening unannotated splice sites. Selecting 'masked' will hide the score for such splicing changes and show 0 instead. Selecting 'raw' will show all scores. SpliceAI developers recommend using 'raw' scores for alternative splicing analysis and 'masked' scores for variant interpretation."></i>
                            </div>
                            <div class="field" style="padding-right:10px; padding-bottom:12px">
                                <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                <div class="ui input" style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:80px">
                                    <input id="max-distance-input" type="text" value="50" style="padding:7px 10px">
                                </div>
                                <i class="question circle outline icon" data-position="right center" data-content="For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>
                            </div>
                        </div>
                        <div class="three wide column">
                            <button id="submit-button" class="ui primary button">Submit</button>
                        </div>
                    </div>
            </div>
            </div>
        </div>
        <div class="eight wide column"></div>
    </div>
        <div class="row">
            <div class="one wide column"></div>
            <div class="fifteen wide column">
                <div>
                    <div id="error-box" style="color:darkred"></div>
                    <div id="response-box"></div>
                </div>
            </div>
        </div>
    </div>

    <script>
      var VARIANT_RE = new RegExp(
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\s:]+" +
        "([0-9]{1,9})" +
        "[-\\s:]+" +
        "([ACGT]+)" +
        "[-\\s:>]+" +
        "([ACGT]+)",
        'i'
      )

      var GRCH37_contigs = {
        "1": "NC_000001.10",
        "2": "NC_000002.11",
        "3": "NC_000003.11",
        "4": "NC_000004.11",
        "5": "NC_000005.9",
        "6": "NC_000006.11",
        "7": "NC_000007.13",
        "8": "NC_000008.10",
        "9": "NC_000009.11",
        "10": "NC_000010.10",
        "11": "NC_000011.9",
        "12": "NC_000012.11",
        "13": "NC_000013.10",
        "14": "NC_000014.8",
        "15": "NC_000015.9",
        "16": "NC_000016.9",
        "17": "NC_000017.10",
        "18": "NC_000018.9",
        "19": "NC_000019.9",
        "20": "NC_000020.10",
        "21": "NC_000021.8",
        "22": "NC_000022.10",
        "X": "NC_000023.10",
        "Y": "NC_000024.9",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      var GRCH38_contigs = {
        "1" : "NC_000001.11",
        "2" : "NC_000002.12",
        "3" : "NC_000003.12",
        "4" : "NC_000004.12",
        "5" : "NC_000005.10",
        "6" : "NC_000006.12",
        "7" : "NC_000007.14",
        "8" : "NC_000008.11",
        "9" : "NC_000009.12",
        "10": "NC_000010.11",
        "11": "NC_000011.10",
        "12": "NC_000012.12",
        "13": "NC_000013.11",
        "14": "NC_000014.9",
        "15": "NC_000015.10",
        "16": "NC_000016.10",
        "17": "NC_000017.11",
        "18": "NC_000018.10",
        "19": "NC_000019.10",
        "20": "NC_000020.11",
        "21": "NC_000021.9",
        "22": "NC_000022.11",
        "X" : "NC_000023.11",
        "Y" : "NC_000024.10",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      var SPLICEAI_SCORE_FIELDS = "SYMBOL|DS_AG|DS_AL|DS_DG|DS_DL|DP_AG|DP_AL|DP_DG|DP_DL".split("|")

      function parseVariant(variant) {
        var m = variant.match(VARIANT_RE)
        if (m) {
          var chrom = m[2].toUpperCase()
          return {
            chrom: chrom,
            pos: m[3],
            ref: m[4],
            alt: m[5],
          }
        }

        return null
      }


      function parseVariantToHGVS(variant, genome_version) {
        var m = variant.match(VARIANT_RE)
        if (m) {
          var chrom = m[2].toUpperCase()
          if (genome_version == "37") {
            chrom = GRCH37_contigs[chrom]
          } else if(genome_version == "38") {
            chrom = GRCH38_contigs[chrom]
          } else {
            throw "Invalid genome_version: " + genome_version
          }

          var pos = parseInt(m[3])
          var ref = m[4]
          var alt = m[5]

          if (ref.length == alt.length) {
            variant = chrom + ":g." + pos + ref + ">" + alt
          } else if (ref.length > alt.length) {
            //deletion (https://varnomen.hgvs.org/recommendations/DNA/variant/deletion/)
            variant = chrom + ":g." + (pos + 1) + "_" + (pos + ref.length - alt.length) + "del"
          } else if (ref.length < alt.length) {
            //insertion (https://varnomen.hgvs.org/recommendations/DNA/variant/insertion/)
            variant = chrom + ":g." + pos + "_" + (pos + 1) + "ins" + alt.slice(1)
          }

          console.log("Parsed variant", variant)
        }

        return variant
      }


      function getScoreStyle(score) {
        var red = "#fccfb8"
        var yellow = "#fff19d"
        var green = "#cdffd7"

        var score = parseFloat(score)
        if(score >= 0.8) {
          return "style='background-color:"+red+"'"
        } else if (score >= 0.5) {
          return "style='background-color:"+yellow+"'"
        } else if (score >= 0.2) {
          return "style='background-color:"+green+"'"
        }
      }

      function getUCSCBrowserUrl(genome_version, chrom, pos) {

        genome_version = genome_version.replace('37', '19')
        chrom = chrom.toUpperCase().replace('CHR', '')
        return "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg" + genome_version + "&position=chr" + chrom + ":" + pos
      }

      function generateResultsTable(response) {
        console.log("Processing result:", response)

        try {
          response["url"] = getUCSCBrowserUrl(response.genome_version, response.chrom, response.pos)
        } catch(e) {
          response["url"] = "#"
        }


        var parsed_scores = response.scores.map(function(s) {
          var splice_ai_values = s.split("|")
          var splice_ai_dict = {}
          SPLICEAI_SCORE_FIELDS.forEach(function(k, i) {
            splice_ai_dict[k] = splice_ai_values[i]
          })

          var score_types = ["DG", "DL", "AG", "AL"]
          score_types.forEach(function(score_type) {
            try {
              splice_ai_dict[score_type + "_url"] = getUCSCBrowserUrl(response.genome_version, response.chrom, response.pos + parseInt(splice_ai_dict["DP_" + score_type]))
            } catch(e) {
              splice_ai_dict[score_type + "_url"] = "#"
            }
          })

          return splice_ai_dict
        })

        console.log("Parsed scores", parsed_scores)

        return "<table>" +
          "<tr>" +
          "<td><b>variant</b></td>" +
          "<td><b>gene</b></td>" +
          "<td><b>&#916;&nbsp; type</b></td>" +
          "<td><b>&#916;&nbsp; score</b>  " + "<i class='question circle outline icon' data-content='Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 50bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red.'></i>"+"</td>" +
          "<td><b>pre-mRNA position</b>  "+ "<i class='question circle outline icon' data-content='For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant.'></i>"+"</td>" +
          "</tr>" + parsed_scores.map(function(scores) {
                return "<tr class='upperBorder'>" +
                  "<td><a href='"+response.url+"' target='_blank'>"+response.variant+"</a></td>"+
                  "<td><a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene="+scores.SYMBOL+"' target='_blank'>"+scores.SYMBOL+"</a></td>" +
                  "<td>Acceptor Loss</td><td "+getScoreStyle(scores.DS_AL)+">"+scores.DS_AL+"</td><td>" + (parseFloat(scores.DS_AL) == 0 ? "" : /*"<a href='"+scores.AL_url+"' target='_blank'>"+ */scores.DP_AL+" bp"/*+</a>"*/) + "</td>"+
                  "</tr><tr><td></td><td></td>"+
                  "<td>Donor Loss</td><td "+getScoreStyle(scores.DS_DL)+">"+scores.DS_DL+"</td><td>" + (parseFloat(scores.DS_DL) == 0 ? "" : /*"<a href='"+scores.DL_url+"' target='_blank'>"+ */scores.DP_DL+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr><tr><td></td><td></td>"+
                  "<td>Acceptor Gain</td><td "+getScoreStyle(scores.DS_AG)+">"+scores.DS_AG+"</td><td>" + (parseFloat(scores.DS_AG) == 0 ? "" : /*"<a href='"+scores.AG_url+"' target='_blank'>"+ */scores.DP_AG+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr><tr><td></td><td></td>"+
                  "<td>Donor Gain</td><td "+getScoreStyle(scores.DS_DG)+">"+scores.DS_DG+"</td><td>" + (parseFloat(scores.DS_DG) == 0 ? "" : /*"<a href='"+scores.DG_url+"' target='_blank'>"+ */scores.DP_DG+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr>"

              }).join("") +
          "</table>"
      }

      $(document).ready(function() {
        $('.ui.radio.checkbox').checkbox()
        $('.question').popup()
        $("#search-box").focus()
        $("#search-box,#max-distance-input").keydown(function (event) {
          if ((event.keyCode || event.which) == 13) {
            $("#submit-button").click()
          }
        });
        $("#submit-button").click(function() {
          var variant = $("#search-box").val().trim()
          if (!variant) {
            return
          }

          var genome_version = $("input[name='hg']:checked").val().trim()
          var mask = $("input[name='mask']:checked").val().trim()
          var max_distance = $("#max-distance-input").val().trim()

          var variant_hgvs = parseVariantToHGVS(variant, genome_version)

          $("#error-box").hide()
          $("#response-box").hide()

          $("#submit-button").addClass(["loading", "disabled"])

          var ensembl_api_url_prefix = ""
          if (genome_version == "37") {
            ensembl_api_url_prefix = "grch37."
          }
          $.getJSON(
            "https://" + ensembl_api_url_prefix + "rest.ensembl.org/vep/human/hgvs/" + variant_hgvs + "?content-type=application/json&vcf_string=1"
          ).catch(function(error) {
            var errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
              console.warn(e)
            }

            var refAlleleError = errorText.match(new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
            if (refAlleleError) {
                errorText = variant + " has unexpected reference allele. The hg" + genome_version + " reference allele should be: " + refAlleleError[1]
            }

            errorText = "Ensembl API error: " + errorText
            console.warn(errorText)
            throw { responseText: errorText }
          }).then(function(response) {
            console.log("Esnembl API response", response)
            if (!response || !response[0]) {
              throw { responseText: "Unable to parse Ensembl API repsonse: " + response }
            }

            var vcf_string = (response[0].vcf_string || "---").split("-")
            var chrom = vcf_string[0]
            var pos = vcf_string[1]
            var ref = vcf_string[2]
            var alt = vcf_string[3]
            var most_severe_consequence = (response[0].most_severe_consequence || "").replace(/_/g, " ")

            console.log("chrom:", chrom, "pos:", pos, "ref:", ref, "alt:", alt)
            console.log("most_severe_consequence:", most_severe_consequence)

            // call SpliceAI-lookup API
            return $.getJSON(
              //"http://localhost:8080/spliceai/",
              "https://spliceailookup-api.broadinstitute.org/spliceai/",
              {
                hg: genome_version,
                distance: max_distance,
                mask: mask,
                variant: chrom+'-'+pos+'-'+ref+'-'+alt,
              }
            )
          }).then(function(response) {
            $("#error-box").hide()
            $("#response-box").html(generateResultsTable(response))
            $("#response-box").show()
          }).catch(function(error) {
            $("#response-box").hide()
            var errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
            }

            if (!errorText) {
                errorText = "ERROR: An unknown error occurred."
            }
            console.warn(errorText)
            $("#error-box").html(errorText)
            $("#error-box").show()
          }).always(function() {
            $("#submit-button").removeClass(["loading", "disabled"]);
            $('.question').popup()
          })
        })
      })
    </script>
</body>
</html>
