<html>
<head>
    <meta content="width=device-width, initial-scale=1" charset="utf-8" name="viewport" />
    <title>SpliceAI Lookup</title>
    <link rel="shortcut icon" href="/icon.png">

    <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.css" />
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/semantic-ui@2.4.2/dist/semantic.min.js"></script>
    <script type="text/javascript" src="https://cdn.jsdelivr.net/npm/underscore@1.12.0/underscore-min.js"></script>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-172722632-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-172722632-1');
    </script>
    <style>
        body {
            overflow-x: scroll;
        }
        body, .ui.form, label {
            font-size: 11pt !important;
        }
        td {
            padding: 5px 15px;
        }
        .upperBorder {
            border-top: 1px solid black;
        }
        table {
            font-size: 11pt !important;
            border-collapse: collapse;
            border-left: 0px !important;
            border-right: 0px !important;
        }
        .smallLink {
            font-size: 10pt;
            margin-left: 5px;
            margin-top: 3px;
        }

        .only-large-screen {
            display: none !important;
        }

        @media screen and (min-width: 770px) {
            .only-large-screen {
                display: block !important;
            }
        }

        .results-div {
            width: 100%;
            position: static;
        }

        @media screen and (max-width: 1400px) {
            .results-div {
                width: 85vw;
                position: relative;
            }
        }

    </style>
</head>
<body>
    <div class="ui stackable grid">
      <div class="row">
        <div class="one wide column only-large-screen"></div>
        <div class="ten wide column">
            <div class="ui grid">
                <div class="row">
                    <div class="twelve wide column">
                        <div style="padding:25px 0px">
                            <div>
                                Enter a variant below to see its <a href="https://www.cell.com/cell/pdf/S0092-8674(18)31629-5.pdf" target="_blank">SpliceAI</a> scores.
                                This can be any SNV or simple InDel within an exon or intron defined by GENCODE v37.<br />
                            </div>
                            <div>
                                <br />
                                <a id="more-details1-button" href="#" onclick="$('#more-details1').show();$('#more-details1-button').hide();">more details...</a>

                                <div id="more-details1" style="display:none">
                                    This web interface and the underlying <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> are being developed by the <a href="https://the-tgg.org/" target="_blank">TGG</a> at the <a href="https://www.broadinstitute.org/" target="_blank">Broad Institute</a>.<br />
                                    The <a href="https://spliceailookup-api.broadinstitute.org/" target="_blank">SpliceAI server</a> optionally uses pre-computed
                                    scores provided by Illumina for SNVs and small InDels, but resorts to running the <a href="https://github.com/Illumina/SpliceAI" target="_blank">SpliceAI model</a>
                                    directly for scores not available in the pre-computed files - such as larger InDels or non-default "Max distance" values.<br />
				    <br />
				    NOTE: The API currently only supports interactive use. For automated batch processing, please use the
				    <a href="https://github.com/broadinstitute/SpliceAI-lookup/blob/master/README.md#local-install" target="_blank">source code</a>
				    to set up your own local instance of the API server or just run the underlying SpliceAI tool.<br />
                                    <br />
                                    For more information on the SpliceAI model, see <a href="https://doi.org/10.1016/j.cell.2018.12.015" target="_blank">Jaganathan et al, Cell 2019</a> and <a href="https://github.com/Illumina/SpliceAI" target="_blank">github.com/Illumina/SpliceAI</a>,
                                    as well as in this <a href="https://www.youtube.com/watch?v=oJvhj-tYbBI" target="_blank">recorded talk</a> by Kishore Jaganathan.<br />
                                    <br />
                                </div>
                            </div>
                            <br />
                            <div style="padding-left: 15px; color: gray">
                                <table>
                                    <thead style="text-align: left">
                                        <tr>
                                            <th style="padding-bottom: 20px">Examples &nbsp; (on hg38): </th>
                                            <th></th>
                                        </tr>
                                    </thead>
                                    <tbody>
                                        <tr>
                                            <td>chr8-140300616-T-G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td>chr8 &nbsp; 140300616 &nbsp; T &nbsp; G</td>
                                            <td></td>
                                        </tr>
                                        <tr>
                                            <td>NM_000552.4(VWF):c.3797C>A (p.Pro1266Gln)</td>
                                            <td></td>
                                        </tr>
                                    </tbody>
                                </table>
                            </div>
                            <br />
                            <a id="more-details2-button" href="#" onclick="$('#more-details2').show();$('#more-details2-button').hide();">more examples...</a>
                            <div id="more-details2" style="display:none; padding-left: 15px; color: gray">
                                <table>
                                    <thead></thead>
                                    <tbody>
                                        <tr>
                                            <td>8:140300616 T&gt;G</td>
                                            <td>'chr' <i>prefix is optional</i></td>
                                        </tr>
                                        <tr>
                                            <td>1:930130:C:G </td>
                                            <td><i>position with overlapping genes</i></td>
                                        </tr>
                                        <tr>
                                            <td>1-1042601-A-AGAGAG </td>
                                            <td><i>insertion of GAGAG</i></td>
                                        </tr>
                                        <tr>
                                            <td>1-1042466-GGGC-G</td>
                                            <td><i>deletion of GGC</i></td>
                                        </tr>
					<tr>
					  <td>NM_000249.4(MLH1):c.116G>A</td>
					  <td><i>another HGVS exapmle</i></td>
					</tr>
                                    </tbody>
                                </table>
                            </div>
                        </div>


                        <br />
                        <div style="font-size: large; color: darkred">
                          July 6, 2021 7:30am EST<br />
                          <br />
                          <b>The server is briefly down for updates</b><br />
                          <br />
                          Please check back in 1 hour
                        </div>
                        <br />
                        <br />


                        <div class="ui form" style="width:100%">
                            <div class="ui input" style="padding-bottom:30px; padding-right:8px; width:100%">
                                <input id="search-box" type="text" style="width: 100%; padding: 7px 10px" placeholder="Enter a variant...">
                            </div>
                            <!-- div class="field" style="padding-bottom: 10px;">
                                <textarea id="search-box2" rows="1" style="width: 100%" placeholder="Enter variant..."></textarea>
                            </div -->
                            <div class="ui stackable grid">
                                <div class="row">
                                    <div class="thirteen wide column">
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Genome version:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="37" /><label>hg19</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="hg" value="38" checked /><label>hg38</label>
                                            </div>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Score type:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="1" /><label>masked</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="mask" value="0" checked /><label>raw</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Splicing changes corresponding to strengthening annotated splice sites and weakening unannotated splice sites are typically much less pathogenic than weakening annotated splice sites and strengthening unannotated splice sites. Selecting 'masked' will hide the score for such splicing changes and show 0 instead. Selecting 'raw' will show all scores. SpliceAI developers recommend using 'raw' scores for alternative splicing analysis and 'masked' scores for variant interpretation."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Max distance:</b></span>
                                            <div class="ui input" style="padding-left:15px; padding-bottom:10px; padding-right:8px; width:100px">
                                                <input id="max-distance-input" type="text" value="50" style="padding:7px 10px">
                                            </div>
                                            <i class="question circle outline icon" data-position="right center" data-content="For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions being splice acceptors or donors. The distance specified here controls the size of this window. The maximum allowed value is 10,000bp."></i>
                                        </div>
                                        <div class="field" style="padding-right:10px; padding-bottom:12px; white-space:nowrap">
                                            <span style="display: inline-block; white-space: nowrap; padding-bottom:10px;"><b>Use Illumina's pre-computed scores:</b></span>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="1" /><label>yes</label>
                                            </div>
                                            <div class="ui radio checkbox" style="padding-left:15px; padding-bottom:10px">
                                                <input type="radio" name="use-precomputed-scores" value="0" checked /><label>no</label>
                                            </div>
                                            <i style="margin-left:20px" class="question circle outline icon" data-position="right center" data-content="Illumina provides pre-computed SpliceAI scores for all SNPs and small InDels within Gencode v24 canonical transcripts. If 'yes' is selected here, the server will quickly check these pre-computed files to see if it can avoid the relatively slow step of running the SpliceAI model. Even with 'yes' selected, it will still run the model if it finds that the pre-computed files don't contain the variant. Selecting 'no' will cause the server to always skip the pre-computed files and run the model. The benefit of skipping the pre-computed files is that, although the scores may take longer, they will always be computed for Gencode v36 transcripts and will include ENSG and ENST ids in addition to the gene name."></i>
                                        </div>
                                    </div>
                                    <div class="three wide column">
                                        <button id="submit-button" class="ui primary button" style="margin-bottom: 15px">Submit</button>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        </div>
        <div class="four wide column">
            <div class="only-large-screen" style="padding:25px 0px">
                <div><i>What's New</i></div>
                <br />
                <div>
                    To post issues or feature requests, please use<br /><a href="https://github.com/broadinstitute/SpliceAI-lookup/issues" target="_blank">SpliceAI-lookup/issues</a><br />
                    <br />
                    <b>June 2, 2021</b><br />
                    - show RefSeq ids for the subset of Ensembl ENST ids (~30%) that have one or more matching RefSeq NM ids according to Ensembl. See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/8">issue #8</a> for details.<br />
		    - update to <a href="https://www.gencodegenes.org/human/releases.html">Gencode v38</a> (was previously on v37)<br />
		    <br />
                    <b>April 11, 2021</b><br />
                    - show gray background for non-coding transcripts<br />
                    - switch to showing all non-coding transcripts. (Previously, transcripts with Gencode biotypes
                      like <i>lncRNA</i>, <i>processed_pseudogene</i>, <i>processed_transcript</i>, <i>retained_intron</i>, and
                      <i>nonsense_mediated_decay</i> were filtered out if they overlapped <i>protein_coding</i> transcripts). 
		      See <a href="https://github.com/broadinstitute/SpliceAI-lookup/issues/6#issuecomment-816942824">issue #6</a> for details.<br />
		    - update to Gencode v37 (was previously on v36)<br />
                </div>
                <br />
                <div>
                    <b>Related web tools:</b><br />
                    <a href="https://liftover.broadinstitute.org">liftover</a> - simple hg19 <=> hg38 conversion for variants, positions, intervals<br />
                    <a href="https://cma-search.broadinstitute.org">CMA search</a> - search OMIM by genomic region, gene name, phenotype, etc.<br />
                </div>
            </div>
        </div>
          <div class="one wide column only-large-screen"></div>
      </div>
      <div class="row">
          <div class="one wide column only-large-screen"></div>
          <div class="fourteen wide column">
              <div class="results-div">
                  <div id="error-box" style="color:darkred"></div>
                  <div id="response-box"></div>
              </div>
          </div>
          <div class="one wide column only-large-screen"></div>
      </div>
    </div>

    <script>
      var VARIANT_RE = new RegExp(
	"^[\\s]*" +
        "(chr)?([0-9XYMTt]{1,2})" +
        "[-\\p{Pd}\\s:]+" +
        "([0-9,]+)" +
        "[-\\p{Pd}\\s:]*" +
        "([ACGT]+)" +
        "[-\\p{Pd}\\s:>]+" +
        "([ACGT]+)",
        'iu'
      )

      var GRCH37_contigs = {
        "1": "NC_000001.10",
        "2": "NC_000002.11",
        "3": "NC_000003.11",
        "4": "NC_000004.11",
        "5": "NC_000005.9",
        "6": "NC_000006.11",
        "7": "NC_000007.13",
        "8": "NC_000008.10",
        "9": "NC_000009.11",
        "10": "NC_000010.10",
        "11": "NC_000011.9",
        "12": "NC_000012.11",
        "13": "NC_000013.10",
        "14": "NC_000014.8",
        "15": "NC_000015.9",
        "16": "NC_000016.9",
        "17": "NC_000017.10",
        "18": "NC_000018.9",
        "19": "NC_000019.9",
        "20": "NC_000020.10",
        "21": "NC_000021.8",
        "22": "NC_000022.10",
        "X": "NC_000023.10",
        "Y": "NC_000024.9",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      var GRCH38_contigs = {
        "1" : "NC_000001.11",
        "2" : "NC_000002.12",
        "3" : "NC_000003.12",
        "4" : "NC_000004.12",
        "5" : "NC_000005.10",
        "6" : "NC_000006.12",
        "7" : "NC_000007.14",
        "8" : "NC_000008.11",
        "9" : "NC_000009.12",
        "10": "NC_000010.11",
        "11": "NC_000011.10",
        "12": "NC_000012.12",
        "13": "NC_000013.11",
        "14": "NC_000014.9",
        "15": "NC_000015.10",
        "16": "NC_000016.10",
        "17": "NC_000017.11",
        "18": "NC_000018.10",
        "19": "NC_000019.10",
        "20": "NC_000020.11",
        "21": "NC_000021.9",
        "22": "NC_000022.11",
        "X" : "NC_000023.11",
        "Y" : "NC_000024.10",
        "M":  "NC_012920.1",
        "MT": "NC_012920.1",
      }

      var SPLICEAI_SCORE_FIELDS = "SYMBOL|DS_AG|DS_AL|DS_DG|DS_DL|DP_AG|DP_AL|DP_DG|DP_DL".split("|")

      function parseVariant(variant) {
        var m = variant.match(VARIANT_RE)
        if (m) {
          var chrom = m[2].toUpperCase()
          return {
            chrom: chrom,
            pos: m[3].replace(/,/g, ""),
            ref: m[4],
            alt: m[5],
          }
        }

        return null
      }


      function parseVariantToHGVS(variant, genome_version) {
        var m = variant.match(VARIANT_RE)
        if (m) {
          var chrom = m[2].toUpperCase()
          if (genome_version == "37") {
            chrom = GRCH37_contigs[chrom]
          } else if(genome_version == "38") {
            chrom = GRCH38_contigs[chrom]
          } else {
            throw "Invalid genome_version: " + genome_version
          }

          var pos = parseInt(m[3].replace(/,/g, ""))
          var ref = m[4]
          var alt = m[5]

          if (ref.length == alt.length) {
            variant = chrom + ":g." + pos + ref + ">" + alt
          } else if (ref.length > alt.length) {
            //deletion (https://varnomen.hgvs.org/recommendations/DNA/variant/deletion/)
            variant = chrom + ":g." + (pos + 1) + "_" + (pos + ref.length - alt.length) + "del"
          } else if (ref.length < alt.length) {
            //insertion (https://varnomen.hgvs.org/recommendations/DNA/variant/insertion/)
            variant = chrom + ":g." + pos + "_" + (pos + 1) + "ins" + alt.slice(1)
          }

          console.log("Parsed variant", variant)
        }

        return variant
      }


      function getScoreStyle(score) {
        var red = "#fccfb8"
        var yellow = "#fff19d"
        var green = "#cdffd7"

        var score = parseFloat(score)
        if(score >= 0.8) {
          return "style='background-color:"+red+"'"
        } else if (score >= 0.5) {
          return "style='background-color:"+yellow+"'"
        } else if (score >= 0.2) {
          return "style='background-color:"+green+"'"
        }
      }

      function getUCSCBrowserUrl(genome_version, chrom, pos) {

        genome_version = genome_version.replace('37', '19')
        chrom = chrom.toUpperCase().replace('CHR', '')
        return "https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg" + genome_version + "&position=chr" + chrom + ":" + pos
      }

      function generateResultsTable(response) {
        console.log("Processing result:", response)

        try {
          response["url"] = getUCSCBrowserUrl(response.hg, response.chrom, response.pos)
        } catch(e) {
          response["url"] = "#"
        }


        var parsed_scores = response.scores.map(function(s) {
          var splice_ai_values = s.split("|")
          var splice_ai_dict = {}
          SPLICEAI_SCORE_FIELDS.forEach(function(k, i) {
            splice_ai_dict[k] = splice_ai_values[i]
          })

          var score_types = ["DG", "DL", "AG", "AL"]
          score_types.forEach(function(score_type) {
            try {
              splice_ai_dict[score_type + "_url"] = getUCSCBrowserUrl(response.hg, response.chrom, response.pos + parseInt(splice_ai_dict["DP_" + score_type]))
            } catch(e) {
              splice_ai_dict[score_type + "_url"] = "#"
            }
          })

          return splice_ai_dict
        })

        console.log("Parsed scores", parsed_scores)

        //sort by (gene name, is canonical transcript, is non-coding transcriipt)
        parsed_scores = _.chain(parsed_scores)
          .sortBy(function(scores) {
            const gene_name_fields = (scores.SYMBOL || "").split("---")
            const gene_name = gene_name_fields[0]
            const transcript_id = gene_name_fields.length >= 3 && gene_name_fields[2]
            const is_canonical_transcript = (gene_name_fields.length >= 4 && gene_name_fields[3] === "yes") ? 0 : 1
            const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
            const is_protein_coding = !transcript_type || transcript_type == "protein_coding" ? 0 : 1
            const refseq_transcript_id_is_in_query = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",").some(refseq_transcript_id => response.raw.includes(refseq_transcript_id))
            const contains_hgvs_transcript_id = response.raw.includes(transcript_id) || refseq_transcript_id_is_in_query ? 0 : 1
            return `${contains_hgvs_transcript_id}${is_protein_coding}${is_canonical_transcript}---${gene_name}`
          })

        return "<table class='ui stackable table'>" +
          "<tr>" +
              "<td><b>variant</b></td>" +
              "<td><b>gene</b></td>" +
              "<td><b>&#916;&nbsp; type</b></td>" +
              "<td><b>&#916;&nbsp; score</b>  " + "<i class='question circle outline icon' data-content='Delta scores range from 0 to 1 and can be interpreted as the probability that the variant affects splicing at any position within a window around it (+/- 50bp by default). In the SpliceAI paper, a detailed characterization is provided for 0.2 (high recall), 0.5 (recommended), and 0.8 (high precision) cutoffs. Consequently, scores of 0.2 or higher appear in green, 0.5 or higher appear in yellow, and 0.8 or higher appear in red.'></i>"+"</td>" +
              "<td><b>pre-mRNA position</b>  "+ "<i class='question circle outline icon' data-content='For each variant, SpliceAI looks within a window (+/- 50bp by default) to see how the variant affects the probabilities of different positions in the pre-mRNA being splice acceptors or donors. The offsets in this column represent positions with the biggest change in probability within the window. Negative values are upstream (5&#039;) of the variant, and positive are downstream (3&#039;) of the variant.'></i>"+"</td>" +
          "</tr>" + parsed_scores.map(function(scores) {
                const variant_ucsc_url = response.url
                const gnomAD_version = response.hg === "38"? "gnomad_r3" : "gnomad_r2_1"
                const variant_gnomAD_url = "https://gnomad.broadinstitute.org/variant/" + response.variant + "?dataset=" + gnomAD_version
                const gene_name_fields = (scores.SYMBOL || "").split("---")
                const gene_name = gene_name_fields[0]

                const is_canonical_transcript = gene_name_fields.length >= 4 && gene_name_fields[3] === "yes"
                const transcript_type = gene_name_fields.length >= 5 && gene_name_fields[4]
                const is_protein_coding = !transcript_type || transcript_type == "protein_coding"
                const refseq_transcript_ids = gene_name_fields.length >= 6 && gene_name_fields[5] && gene_name_fields[5].split(",")
                const background_color = !is_protein_coding ? "#e5e5e5" : (is_canonical_transcript ? "#eaf3ff" : "white")

                let gene_and_transcript_id = ""
                if (gene_name_fields.length >= 3 ) {
                    const gene_id_link = `<a href="https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g=${gene_name_fields[1].split(".")[0]}" target="_blank">${gene_name_fields[1]}</a>`
                    const transcript_id_link = ` / <a href="https://useast.ensembl.org/Homo_sapiens/Transcript/Summary?t=${gene_name_fields[2].split(".")[0]}" target="_blank">${gene_name_fields[2]}</a>`
                    const canonical_transcript_text = is_canonical_transcript ? '<div class="smallLink">canonical transcript</div>' : ''
                    const biotype_link = `<div class="smallLink"><a href="https://www.gencodegenes.org/pages/biotypes.html" target="_blank">biotype: ${transcript_type.replace(/_/g, " ")}</a></div>`
                    const refseq_transcript_id_link = refseq_transcript_ids ? (" /" + refseq_transcript_ids.map(refseq_transcript_id => ` <a href="https://www.ncbi.nlm.nih.gov/search/all/?term=${refseq_transcript_id}" target="_blank">${refseq_transcript_id}</a>`).join(', ')) : ''
                    gene_and_transcript_id = ` (${gene_id_link}${transcript_id_link}${refseq_transcript_id_link})<br /> ${biotype_link} ${canonical_transcript_text}`
                }

                return "<tr class='upperBorder' style='background-color:" + background_color+"'>" +
                  "<td rowspan='4' style='vertical-align: top'>" +
                  "<div style='margin-right:10px; display: inline-block'>" + response.variant + "</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='"+variant_ucsc_url+"' class='smallLink' target='_blank'>UCSC</a>, " +
                    "<a href='"+variant_gnomAD_url+"' class='smallLink' target='_blank'>gnomAD</a>" +
                  "</td>"+
                  "<td rowspan='4' style='vertical-align: top'>" +
                    "<div style='margin-right:10px; display: inline-block'>"+gene_name+gene_and_transcript_id+"</div><br class='only-large-screen'/><br class='only-large-screen'/>" +
                    "<a href='https://www.omim.org/search?search="+gene_name+"' class='smallLink' target='_blank'>OMIM</a>, " +
                    "<a href='https://gtexportal.org/home/gene/"+gene_name+"' class='smallLink' target='_blank'>GTEx</a>, " +
                    "<a href='https://gnomad.broadinstitute.org/gene/"+gene_name+"?dataset="+gnomAD_version+"' class='smallLink' target='_blank'>gnomAD</a>, " +
                    "<a href='https://search.clinicalgenome.org/kb/genes?page=1&size=25&search="+gene_name+"' class='smallLink' target='_blank'>ClinGen</a>, <br class='only-large-screen'/>" +
                    "<a href='https://useast.ensembl.org/Homo_sapiens/Gene/Summary?g="+gene_name+"' class='smallLink' target='_blank'>Ensembl</a>, " +
                    "<a href='https://decipher.sanger.ac.uk/gene/"+gene_name+"' class='smallLink' target='_blank'>Decipher</a>, " +
                    "<a href='https://www.genecards.org/cgi-bin/carddisp.pl?gene="+gene_name+"' class='smallLink' target='_blank'>GeneCards</a>" +
                  "</td>" +
                  "<td>Acceptor Loss</td><td "+getScoreStyle(scores.DS_AL)+">"+scores.DS_AL+"</td><td>" + (parseFloat(scores.DS_AL) == 0 ? "" : /*"<a href='"+scores.AL_url+"' target='_blank'>"+ */scores.DP_AL+" bp"/*+</a>"*/) + "</td>"+
                  "</tr><tr style='background-color:" + background_color+"'>"+
                  "<td>Donor Loss</td><td "+getScoreStyle(scores.DS_DL)+">"+scores.DS_DL+"</td><td>" + (parseFloat(scores.DS_DL) == 0 ? "" : /*"<a href='"+scores.DL_url+"' target='_blank'>"+ */scores.DP_DL+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr><tr style='background-color:" + background_color+"'>"+
                  "<td>Acceptor Gain</td><td "+getScoreStyle(scores.DS_AG)+">"+scores.DS_AG+"</td><td>" + (parseFloat(scores.DS_AG) == 0 ? "" : /*"<a href='"+scores.AG_url+"' target='_blank'>"+ */scores.DP_AG+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr><tr style='background-color:" + background_color+"'>"+
                  "<td>Donor Gain</td><td "+getScoreStyle(scores.DS_DG)+">"+scores.DS_DG+"</td><td>" + (parseFloat(scores.DS_DG) == 0 ? "" : /*"<a href='"+scores.DG_url+"' target='_blank'>"+ */scores.DP_DG+" bp"/*+"</a>"*/) + "</td>"+
                  "</tr>"

              }).join("") +
          "</table>"
      }

      $.urlParam = function (name) {
        var results = new RegExp('[\?&#]?' + name + '=([^&#]*)').exec(window.location.hash);
        return (results !== null) ? decodeURIComponent(results[1]) || 0 : false;
      }

      $(document).ready(function() {
        $('.ui.radio.checkbox').checkbox()
        $('.question').popup()
        $("#search-box").focus()

        $("#search-box,#max-distance-input").keydown(function (event) {
          if ((event.keyCode || event.which) == 13) {
            $("#submit-button").click()
          }
        });

        $("#submit-button").click(function() {
          var variant = $("#search-box").val().trim()
          if (!variant) {
            return
          }

          var genome_version = $("input[name='hg']:checked").val().trim()
          var mask = $("input[name='mask']:checked").val().trim()
          var max_distance = $("#max-distance-input").val().trim()
          var use_precomputed_scores = $("input[name='use-precomputed-scores']:checked").val().trim()
          var variant_hgvs = parseVariantToHGVS(variant, genome_version)

          $("#error-box").hide()
          $("#response-box").hide()

          $("#submit-button").addClass(["loading", "disabled"])

          var ensembl_api_url_prefix = ""
          if (genome_version == "37") {
            ensembl_api_url_prefix = "grch37."
          }
          $.getJSON(
            "https://" + ensembl_api_url_prefix + "rest.ensembl.org/vep/human/hgvs/" + variant_hgvs + "?content-type=application/json&vcf_string=1"
          ).catch(function(error) {
            var errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
              console.warn(e)
            }

            var refAlleleError = errorText.match(new RegExp("[(]([ACGTRYSWKMBDHVN]+)[)] does not match reference allele given by HGVS notation"))
            if (refAlleleError) {
                errorText = variant + " has unexpected reference allele. The hg" + genome_version + " reference allele should be: " + refAlleleError[1]
            }

            errorText = "Ensembl API error: " + errorText
            console.warn(errorText)
            throw { responseText: errorText }
          }).then(function(response) {
            console.log("Esnembl API response", response)
            if (!response || !response[0]) {
              throw { responseText: "Unable to parse Ensembl API repsonse: " + response }
            }

            var vcf_string = (response[0].vcf_string || "---").split("-")
            var chrom = vcf_string[0]
            var pos = vcf_string[1]
            var ref = vcf_string[2]
            var alt = vcf_string[3]
            var most_severe_consequence = (response[0].most_severe_consequence || "").replace(/_/g, " ")

            console.log("chrom:", chrom, "pos:", pos, "ref:", ref, "alt:", alt)
            console.log("most_severe_consequence:", most_severe_consequence)

            // call SpliceAI-lookup API
            return $.getJSON(
              //"http://localhost:8080/spliceai/",
              "https://spliceailookup-api.broadinstitute.org/spliceai/",
              {
                hg: genome_version,
                distance: max_distance,
                mask: mask,
                precomputed: use_precomputed_scores,
                variant: chrom+'-'+pos+'-'+ref+'-'+alt,
                raw: variant,
              }
            )
          }).then(function(response) {
            $("#error-box").hide()
            $("#response-box").html(generateResultsTable(response))
            $("#response-box").show()

            // set url
            window.location.hash = "#" + $.param({
              variant: variant,
              hg: genome_version,
              distance: max_distance,
              mask: mask,
              precomputed: use_precomputed_scores,
            })

          }).catch(function(error) {
            $("#response-box").hide()
            var errorText = error.responseText
            try {
              errorText = JSON.parse(errorText).error
            } catch(e) {
            }

            if (!errorText) {
                errorText = "ERROR: Unable to reach the SpliceAI-lookup API server. Either the API server is down or your internet isn't working."
            }
            console.warn(errorText)
            $("#error-box").html(errorText)
            $("#error-box").show()
          }).always(function() {
            $("#submit-button").removeClass(["loading", "disabled"]);
            $('.question').popup()
          })
        })

        const hgFromUrl = $.urlParam('hg')
        if (hgFromUrl) {
          $("input[name='hg'][value='"+hgFromUrl+"']").prop("checked", true)
        }

        const maxDistanceFromUrl = $.urlParam('distance')
        if (maxDistanceFromUrl) {
            $("#max-distance-input").val(maxDistanceFromUrl)
        }

        const maskFromUrl = $.urlParam('mask')
        if (maskFromUrl) {
          $("input[name='mask'][value='"+maskFromUrl+"']").prop("checked", true)
        }

        const precomputedFromUrl = $.urlParam('precomputed')
        if (precomputedFromUrl) {
          $("input[name='use-precomputed-scores'][value='"+precomputedFromUrl+"']").prop("checked", true)
        }

        const variantFromUrl = $.urlParam('variant')
        if (variantFromUrl) {
          $("#search-box").val(variantFromUrl)
          $("#submit-button").click()
        }
      })

    </script>
</body>
</html>
